staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,454 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	"k8s.io/apimachinery/pkg/util/httpstream/wsstream"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	constants "k8s.io/apimachinery/pkg/util/portforward"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/klog/v2"
+)
+
+// TunnelingHandler is a handler which tunnels SPDY through WebSockets.
+type TunnelingHandler struct {
+	// Used to communicate between upstream SPDY and downstream tunnel.
+	upgradeHandler http.Handler
+}
+
+// NewTunnelingHandler is used to create the tunnel between an upstream
+// SPDY connection and a downstream tunneling connection through the stored
+// UpgradeAwareProxy.
+func NewTunnelingHandler(upgradeHandler http.Handler) *TunnelingHandler {
+	return &TunnelingHandler{upgradeHandler: upgradeHandler}
+}
+
+// ServeHTTP uses the upgradeHandler to tunnel between a downstream tunneling
+// connection and an upstream SPDY connection. The tunneling connection is
+// a wrapped WebSockets connection which communicates SPDY framed data.
+func (h *TunnelingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	klog.V(4).Infoln("TunnelingHandler ServeHTTP")
+
+	spdyProtocols := spdyProtocolsFromWebsocketProtocols(req)
+	if len(spdyProtocols) == 0 {
+		http.Error(w, "unable to upgrade: no tunneling spdy protocols provided", http.StatusBadRequest)
+		return
+	}
+
+	spdyRequest := createSPDYRequest(req, spdyProtocols...)
+
+	writer := &tunnelingResponseWriter{
+		w: w,
+		conn: &headerInterceptingConn{
+			initializableConn: &tunnelingWebsocketUpgraderConn{
+				w:   w,
```

Comment:
this `w http.ResponseWriter` is passed around many times is hard to think about it, but it seems meanwhile `w: w` and `conn` here are thread safe we are safe